### PR TITLE
T23800

### PR DESCRIPTION
**Pull request by tchx84**
https://phabricator.endlessm.com/T23800

### CODE REVIEW

File: game-state-service
Comment (ptomato) [inline game-state-service]: I think this can just be a subclass of `object`?

File: game-state-service
Comment (ptomato) [inline game-state-service]: use `GLib.SOURCE_REMOVE` for readability

File: game-state-service
Comment (ptomato) [inline game-state-service]: I'm not a big fan of `except: pass`, can you narrow this down to `IOError` or whatever?

File: game-state-service
Comment (ptomato) [inline game-state-service]: I think you can just delete this `except: pass`

File: game-state-service
Comment (ptomato) [inline game-state-service]: Uh oh, why is this necessary? Are you avoiding some kind of race condition?

File: game-state-service
Comment (ptomato) [inline game-state-service]: Is there no need to unpack the variant like in GJS?

File: game-state-service
Comment (ptomato) [inline game-state-service]: Could you combine the two classes into one, or is there a good reason not to do that?

File: game-state-service
Comment (ptomato) [inline game-state-service]: This is supposed to be `GLib.MainLoop`, I think! Assuming this works, is it some sort of pygobject override? Also, I looked at the documentation of `dbus.mainloop.glib.DBusGMainLoop` and I don't understand how it integrates with this.

File: meson.build
Comment (ptomato) [inline meson.build]: Probably good to add `version: '0.0.0'` here or something

File: meson.build
Comment (ptomato) [inline meson.build:24]: at the end of the Meson run it's useful to print out the detected value of `prefix`, `bindir`, and `session_bus_services_dir` for the logs.

**Review by ptomato — CHANGES_REQUESTED** [review]
I'm a bit surprised why you used Python for this when everything else is written in GJS, but it's probably not worth spending time to rewrite it right now. I think in the future we want to avoid context switching as much as we can though.

File: game-state-service
Comment (cosimoc) [inline game-state-service]: Using dbus-glib and dbus-python is a bit of a red flag for me. We should use GIO and GDBus for all new services we build.

File: game-state-service
Comment (cosimoc) [reply game-state-service]: I think there are two other aspects to this pattern:
- we should defer writing the new state file to an idle or a short timeout. That way, if someone calls `set()` in a tight loop, we only write once at the end
- since you use the changes-done hint to reload the state from file, there should be no need to stop the file monitor while you write. But in fact there should be no need to have a file monitor at all, since all the writes go through the DBus service and you already update the state upon a call to `set()`

File: game-state-service
Comment (cosimoc) [inline game-state-service]: What is the interface we're using? Is it documented anywhere? Do we not need a signal to notify when the state has changed?

**Review by cosimoc — CHANGES_REQUESTED** [review]

File: game-state-service
Comment (ptomato) [reply game-state-service]: I understood from this morning's discussion that the idea behind the file monitor is that it's easily hackable by editing the text file.

Comment (cosimoc) [reply game-state-service]: For development or is that something that we'd expose to users?

Comment (ptomato) [reply game-state-service]: In any case for development, and also as something that curious users could discover.

Comment (tchx84) [reply game-state-service]: > Are you avoiding some kind of race condition?

Yes. Otherwise, it would always catch it owns writes. We don't want that.

> we should defer writing the new state file to an idle or a short timeout

I considered this, but when mixed with the "just edit the file" feature, I though it could turned into a tricky game. But I can give it a try.

> But in fact there should be no need to have a file monitor at all

There is, the "just edit the file" feature, is something we want to easily play with scenarios, and interact with the game state.

> For development or is that something that we'd expose to users?

For development really, but like @ptomato said, it would be cool if they find it and hack it.

Comment (cosimoc) [reply game-state-service]: To be honest I don't really buy the "curious user" as a reason to have this...
About development, if it makes the code much more fragile, I do wonder if it's worth it.
I believe that it would be with my suggestion you could avoid the timeout, but on the other hand, if this is a service, it should also quit after a period of inactivity (which it doesn't do right now), and re-activate itself and recreate the state when receiving a new DBus call. So a file monitor is not really worth it, because the service is not supposed to stay around for very long anyway...

File: game-state-service
Comment (tchx84) [reply game-state-service]: There is, .unpack. I am using it now, since I switched to GDBus API.

File: game-state-service
Comment (tchx84) [reply game-state-service]: It makes it a lot easier _for me_ to read and maintain it, specially now with the GDBus API which is more verbose.

File: game-state-service
Comment (tchx84) [reply game-state-service]: Well, I haven't seen any python dbus or GDBus (using python) example using something different from `GObject.MainLoop()`, but `GLib.MainLoop()` seems to work as well.

File: game-state-service
Comment (tchx84) [reply game-state-service]: Switched to GDBus now, hopefully I am using it right, because it seems a bit more verbose.

File: game-state-service
Comment (tchx84) [reply game-state-service]: The iface definition is explicit now, since I am using GDBus.

I think we could have a signal to let clients know that the state changed (globally), but I rather not try to mimic GObject properties-like API (for specific keys) or something like that for now. Since we agreed this Set/Get API was enough for now.

**Comment by tchx84** [comment]
@cosimoc @ptomato addressed all comments now, two specific things I will need your feedback on:

* `changed` signal. To keep it simple, since I am not exposing keys as if they were individual properties, it basically multi-casts the changes to the state.

* `Get` method. To keep this service as flexible as possible, the Set/Get receive/send variants. The thing is, is easy to unpack a receiving Variant, but not the opposite operation (once the object is stored, we lost its signature info). I tried writing a simple guess_signature method, but that got bad early-reviews, so now I just added the signature to the `Get` method (since it should be known anyway). Another alternative would be to store the signature in the state file, but that puts some extra responsibility on the file to not be modified manually.

File: examples.py
Comment (ptomato) [inline examples.py:1]: I think pygobject requires you to specify a version for gi.repository modules nowadays, or you get a warning. `gi.require_version('GLib', '2.0')` (check this in other files as well)

File: game-state-service
Comment (ptomato) [inline game-state-service]: This doesn't need to be a method; make it a constant.

File: game-state-service
Comment (ptomato) [inline game-state-service]: `bytes` is unnecessary here, `encode()` already returns a `bytes`.

File: game-state-service
Comment (cosimoc) [inline game-state-service]: It should be possible to use GApplication (not GtkApplication, but the GLib base class) and simplify the DBus registration code. In particular, GApplication has a built-in "inactivity timeout" mechanism that will make the service quit when inactive for a given amount of time. We want that for this service.

File: game-state-service
Comment (ptomato) [inline game-state-service]: This is also done in the only caller of `_state_to_file()`, so it's redundant.

I think it would be better to remove this one and keep the other one, so that the reset of `_state_save_deferred_id` is next to `SOURCE_REMOVE`.

File: game-state-service
Comment (cosimoc) [inline game-state-service:80]: It may be worth considering splitting this in a real XML file. If you do that, and use gdbus-codegen to generate skeletons for the interface, the code to register objects would get simpler.
Optional for this PR if it's too much churn though.

File: game-state-service
Comment (ptomato) [inline game-state-service]: You can probably make this less verbose by inheriting from `Gio.Application` and implementing the `do_dbus_register` and `do_dbus_unregister` virtual methods.

File: game-state-service
Comment (ptomato) [inline game-state-service]: It's a bit strange to have these format strings up here away from the format parameters, when they're only used in one place anyway. I'll leave that up to you whether you want to change that or keep all the strings in the same place.

**Review by ptomato — CHANGES_REQUESTED** [review]

**Comment by ptomato** [comment]
> @cosimoc @ptomato addressed all comments now, two specific things I will need your feedback on:
> 
> * `changed` signal. To keep it simple, since I am not exposing keys as if they were individual properties, it basically multi-casts the changes to the state.

Sounds OK.

> * `Get` method. To keep this service as flexible as possible, the Set/Get receive/send variants. The thing is, is easy to unpack a receiving Variant, but not the opposite operation (once the object is stored, we lost its signature info). I tried writing a simple guess_signature method, but that got bad early-reviews, so now I just added the signature to the `Get` method (since it should be known anyway). Another alternative would be to store the signature in the state file, but that puts some extra responsibility on the file to not be modified manually.

Also sounds OK.

File: game-state-service
Comment (ptomato) [reply game-state-service]: We should make a decision one way or the other on the "curious user" use case before merging this.

**Review by cosimoc — CHANGES_REQUESTED** [review]
Me and @tchx84 discussed some improvements to the DBus interface to avoid the need to specify a signature when calling `Get()`. Other than that, this looks great overall and I only have a few comments here.

File: game-state-service
Comment (cosimoc) [reply game-state-service]: I don't think that's an interesting use case for now. Let's proceed without and we can revisit later if we want to bring it back.

File: game-state-service
Comment (tchx84) [reply game-state-service]: This is gone now, since I am only using JSON-GLIb API.

File: game-state-service
Comment (tchx84) [reply game-state-service:80]: Ill create a separate issue for this.

**Comment by tchx84** [comment]
@ptomato @cosimoc Now using JSON-GLib and Gio.Application, I think we got a cleaner API and the service behavior we wanted.

```
Set (key, variant)
Get (key) => variant
```

File: game-state-service
Comment (cosimoc) [inline game-state-service]: Should be `do_dbus_unregister()`

File: game-state-service
Comment (cosimoc) [inline game-state-service:133]: You should probably check for `self._dbus_id is not None` before calling into this.

File: game-state-service
Comment (cosimoc) [inline game-state-service:69]: This does not look right; I expect a direct call to `save()` to bypass any deferred work and save, not save only if there is already a deferred work pending.

File: game-state-service
Comment (cosimoc) [inline game-state-service]: Can you set these at construction time?

**Review by cosimoc — CHANGES_REQUESTED** [review]

File: game-state-service
Comment (tchx84) [reply game-state-service:69]: It was a cheap optimization, basically, if there no pending save it means there is really nothing new to save (so we don't write to disk for nothing).

Maybe I should just rename it to "save_if_ needed"? hah

Comment (cosimoc) [reply game-state-service:69]: I see. You could rename the method to something like `flush()`.

File: game-state-service
Comment (tchx84) [reply game-state-service]: U.u

**Pull request merged**
